### PR TITLE
Allow integral moments to be taken against polynomials

### PR DESCRIPTION
**Pull request by mscroggs**
This change allows integral moments to be taken against sets of polynomials as well as against basis functions of other finite elements. Currently Legendre and Chebyshev polynomials are added.

This PR adds two new variants of Lagrange spaces: `integral_legendre` and `integral_chebyshev`. These use integral moments against the polynomials as their DOFs on edges, faces and volumes. These new variants have better scaling than equispaced points, but not as good as GLL or Chebyshev points:

interval-all

In a future PR, these will be used to provide variants of other spaces such as Nedelec and serendipity (progress towards #72)

### CODE REVIEW

File: cpp/basix/e-lagrange.cpp [hunk 1]
```diff
@@ -874,6 +876,140 @@ FiniteElement create_vtk_element(cell::type celltype, int degree,
                        maps::type::identity, discontinuous, degree, degree);
 }
 //-----------------------------------------------------------------------------
+int get_integral_degree(cell::type et, int degree)
```
Comment (garth-wells) [inline cpp/basix/e-lagrange.cpp:879]: `et` is a bit cryptic.

File: cpp/basix/e-lagrange.cpp [hunk 2]
```diff
@@ -874,6 +876,140 @@ FiniteElement create_vtk_element(cell::type celltype, int degree,
...
+  // Create points at nodes, ordered by topology (vertices first)
+  if (degree == 0)
+  {
+    if (!discontinuous)
+    {
+      throw std::runtime_error(
+          "Cannot create a continuous order 0 Lagrange basis function");
+    }
+    auto pt = lattice::create(celltype, 0, lattice::type::equispaced, true);
+    x[tdim].push_back(pt);
```
Comment (garth-wells) [inline cpp/basix/e-lagrange.cpp:933]: Would
```
x[tdim] = {lattice::create(celltype, 0, lattice::type::equispaced, true)};
```
be clearer? The push back makes be expect more than one insertion.

File: cpp/basix/e-lagrange.cpp [hunk 3]
```diff
@@ -874,6 +876,140 @@ FiniteElement create_vtk_element(cell::type celltype, int degree,
...
+    xt::view(M[tdim][0], xt::all(), 0, xt::all()) = xt::eye<double>(num_dofs);
+  }
+  else
+  {
+    for (std::size_t dim = 0; dim <= tdim; ++dim)
+    {
+      M[dim].resize(topology[dim].size());
+      x[dim].resize(topology[dim].size());
+
+      // Point evaluations at vertices
```
Comment (garth-wells) [inline cpp/basix/e-lagrange.cpp:946]: Should this comment be inside the `if`?

File: cpp/basix/e-lagrange.cpp [hunk 4]
```diff
@@ -874,6 +876,140 @@ FiniteElement create_vtk_element(cell::type celltype, int degree,
...
+              = cell::sub_entity_geometry(celltype, dim, e);
+          x[dim][e] = entity_x;
+          const std::size_t num_dofs = entity_x.shape(0);
+          M[dim][e] = xt::xtensor<double, 3>(
+              {num_dofs, static_cast<std::size_t>(1), num_dofs});
+          xt::view(M[dim][e], xt::all(), 0, xt::all())
+              = xt::eye<double>(num_dofs);
+        }
+      }
+      // Integral moments on other subentities
```
Comment (garth-wells) [inline cpp/basix/e-lagrange.cpp:961]: Move inside `else if`.

File: cpp/basix/e-lagrange.cpp [hunk 5]
```diff
@@ -874,6 +876,140 @@ FiniteElement create_vtk_element(cell::type celltype, int degree,
...
+      else if (dim <= tdim)
+      {
+        // This will not work on pyramids and prisms
+        cell::type ct = cell::sub_entity_type(celltype, dim, 0);
+
+        /// TODO: this will need changing for other polynomial types
+        int sub_degree = get_integral_degree(ct, degree);
+
+        if (sub_degree >= 0)
```
Comment (garth-wells) [inline cpp/basix/e-lagrange.cpp:970]:
```cpp
if (int sub_degree = get_integral_degree(ct, degree); sub_degree >= 0)
```

File: cpp/basix/e-lagrange.cpp [hunk 6]
```diff
@@ -874,6 +876,140 @@ FiniteElement create_vtk_element(cell::type celltype, int degree,
...
+    }
+  }
+
+  if (discontinuous)
+  {
+    std::tie(x, M, entity_transformations)
+        = element::make_discontinuous(x, M, entity_transformations, tdim, 1);
+  }
+
+  // auto tensor_factors
```
Comment (garth-wells) [inline cpp/basix/e-lagrange.cpp:1004]: Delete

File: cpp/basix/e-lagrange.cpp [hunk 7]
```diff
@@ -874,6 +876,140 @@ FiniteElement create_vtk_element(cell::type celltype, int degree,
...
+  //    = create_tensor_product_factors(celltype, degree, variant);
+
+  return FiniteElement(element::family::P, celltype, degree, {1},
+                       xt::eye<double>(ndofs), entity_transformations, x, M,
+                       maps::type::identity, discontinuous, degree, degree);
+  //                     tensor_factors);
```
Comment (garth-wells) [inline cpp/basix/e-lagrange.cpp:1010]: delete

File: cpp/basix/e-lagrange.cpp [hunk 8]
```diff
@@ -901,6 +1037,10 @@ FiniteElement basix::element::create_lagrange(cell::type celltype, int degree,
   if (variant == element::lagrange_variant::vtk)
     return create_vtk_element(celltype, degree, discontinuous);
 
+  if (variant == element::lagrange_variant::integral_legendre
+      or variant == element::lagrange_variant::integral_chebyshev)
+    return create_integral_lagrange(celltype, degree, variant, discontinuous);
```
Comment (garth-wells) [inline cpp/basix/e-lagrange.cpp:1042]: Add curly braces for multi-line if.

**Review by garth-wells — CHANGES_REQUESTED** [review]
Some preliminary suggestions.

File: cpp/basix/moments.cpp [hunk 9]
```diff
@@ -226,6 +200,101 @@ xt::xtensor<double, 3> moments::create_dot_moment_dof_transformations(
...
+    for (std::size_t j = 0; j < dim; ++j)
+      for (std::size_t k = 0; k < dim; ++k)
+      {
+        lcoeffs(j, k) = 0;
+        for (std::size_t p = 0; p < pts.shape(0); ++p)
+          lcoeffs(j, k) += wts[p] * lpoly_at_pts(p, j) * poly_at_pts(p, k);
+      }
+
+    for (std::size_t j = 0; j < dim; ++j)
+      for (std::size_t k = 0; k < dim; ++k)
```
Comment (garth-wells) [inline cpp/basix/moments.cpp:286]: Add curly braces

File: cpp/basix/polynomials.cpp [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+}
+//-----------------------------------------------------------------------------
+xt::xtensor<double, 2> chebyshev_quad(int d, const xt::xarray<double>& x)
+{
+  xt::xtensor<double, 2> r0 = chebyshev_interval(d, xt::view(x, xt::all(), 0));
+  xt::xtensor<double, 2> r1 = chebyshev_interval(d, xt::view(x, xt::all(), 1));
+  xt::xtensor<double, 2> out({x.shape(0), static_cast<std::size_t>(polyset::dim(
+                                              cell::type::quadrilateral, d))});
+  std::size_t n = 0;
+  for (std::size_t i = 0; i < r0.shape(1); ++i)
```
Comment (garth-wells) [inline cpp/basix/polynomials.cpp:32]: Add curly braces

File: cpp/basix/polynomials.cpp [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+//-----------------------------------------------------------------------------
+xt::xtensor<double, 2> chebyshev_hex(int d, const xt::xarray<double>& x)
+{
+  xt::xtensor<double, 2> r0 = chebyshev_interval(d, xt::view(x, xt::all(), 0));
+  xt::xtensor<double, 2> r1 = chebyshev_interval(d, xt::view(x, xt::all(), 1));
+  xt::xtensor<double, 2> r2 = chebyshev_interval(d, xt::view(x, xt::all(), 2));
+  xt::xtensor<double, 2> out({x.shape(0), static_cast<std::size_t>(polyset::dim(
+                                              cell::type::hexahedron, d))});
+  std::size_t n = 0;
+  for (std::size_t i = 0; i < r0.shape(1); ++i)
```
Comment (garth-wells) [inline cpp/basix/polynomials.cpp:52]: Add curly braces (2x)

File: cpp/basix/polynomials.cpp [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+  case polynomials::type::chebyshev:
+    return tabulate_chebyshev(celltype, d, x);
+  default:
+    throw std::runtime_error("not implemented yet");
+  }
+}
+//-----------------------------------------------------------------------------
+int polynomials::dim(polynomials::type polytype, cell::type cell, int d)
+{
+  switch (polytype)
```
Comment (garth-wells) [inline cpp/basix/polynomials.cpp:101]: Why is a switch needed here?

**Review by garth-wells — CHANGES_REQUESTED** [review]

File: cpp/basix/polynomials.cpp (hunk 12 above)
Comment (mscroggs) [reply cpp/basix/polynomials.cpp:101]: I was envisaging needing a switch for some other polynomial types (eg bubble polynomials) but as they're not implemented yet, I've removed the switch for now

**Pull request merged**